portfolio/src/main/java/com/google/sps/servlets/VoteServlet.java

```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/vote-results")
+public class VoteServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String poll = request.getParameter("poll");
+
+    Filter pollFilter = new FilterPredicate("poll", FilterOperator.EQUAL, poll);
+    Query query = new Query("Ballot").setFilter(pollFilter);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+
+    // Count the number of votes for each category to minimize JSON reponse.
+    Map<String, Long> categoryVotes = 
+        results.stream()
+               .map(e -> (String) e.getProperty("vote"))
+               .collect(Collectors.groupingBy(str -> str, Collectors.counting()));
+    
+    response.setContentType("application/json");
+    Gson gson = new Gson();
+    String json = gson.toJson(categoryVotes);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String vote = request.getParameter("vote");
+    String poll = request.getParameter("poll");
+
+    Entity ballotEntity = new Entity("Ballot");
+    ballotEntity.setProperty("poll", poll);
+    ballotEntity.setProperty("vote", vote);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(ballotEntity);
+
+    response.sendRedirect("/vote.html");
```

Comment:
nit: Might be nice to redirect them back to the same page they're currently on, so they can see visually how their vote affected the poll

Reply:
That is something that never crossed my mind, but is definitely the right thing to do. Really appreciate this feedback! A fix has been pushed.